service/src/main/java/greencity/service/EventServiceImpl.java

```diff
@@ -137,14 +127,105 @@ public PageableAdvancedDto<EventDto> getAll(Pageable page, Principal principal)
+                .getAddress()).getLatitude() != 0
+                && Objects.requireNonNull(event.getDates().get(event.getDates().size() - 1)
+                    .getAddress()).getLongitude() != 0)
+            .sorted(getComparatorByDates())
+            .collect(Collectors.toList());
+    }
+
+    private List<Event> getOfflineUserEventsSortedCloserToUserLocation(
+        User attender, String userLatitude, String userLongitude) {
+        List<Event> eventsFurtherSorted = getOfflineUserEventsSortedByDate(attender).stream()
+            .filter(event -> LocalDate.now().isBefore(event.getDates().get(event.getDates().size() - 1)
+                .getFinishDate().toLocalDate())
+                || LocalDate.now().isEqual(event.getDates().get(event.getDates().size() - 1)
+                    .getFinishDate().toLocalDate()))
+            .sorted(getComparatorByDistance(Double.parseDouble(userLatitude), Double.parseDouble(userLongitude)))
+            .collect(Collectors.toList());
+        List<Event> eventsPassed = getOfflineUserEventsSortedByDate(attender).stream()
+            .filter(event -> LocalDate.now().isAfter(event.getDates().get(event.getDates().size() - 1)
+                .getFinishDate().toLocalDate()))
+            .collect(Collectors.toList());
+        eventsFurtherSorted.addAll(eventsPassed);
+        return eventsFurtherSorted;
+    }
+
+    private Comparator<Event> getComparatorByDistance(final double userLatitude, final double userLongitude) {
+        return (e1, e2) -> {
+            double distance1 = calculateDistance(userLatitude, userLongitude,
+                Objects.requireNonNull(e1.getDates().get(e1.getDates().size() - 1)
+                    .getAddress()).getLatitude(),
+                Objects.requireNonNull(e1.getDates().get(e1.getDates().size() - 1)
+                    .getAddress()).getLongitude());
+            double distance2 = calculateDistance(userLatitude, userLongitude,
+                Objects.requireNonNull(e2.getDates().get(e2.getDates().size() - 1)
+                    .getAddress()).getLatitude(),
+                Objects.requireNonNull(e2.getDates().get(e2.getDates().size() - 1)
+                    .getAddress()).getLongitude());
+            return Double.compare(distance1, distance2);
+        };
+    }
+
+    private static Comparator<Event> getComparatorByDates() {
+        return (e1, e2) -> e2.getDates().get(e2.getDates().size() - 1).getStartDate()
+            .compareTo(e1.getDates().get(e1.getDates().size() - 1).getStartDate());
+    }
+
+    private static double calculateDistance(double latitude1, double longitude1, double latitude2, double longitude2) {
+        double earthRadius = 6371.0;
+
+        double latitude1Rad = Math.toRadians(latitude1);
+        double longitude1Rad = Math.toRadians(longitude1);
+        double latitude2Rad = Math.toRadians(latitude2);
+        double longitude2Rad = Math.toRadians(longitude2);
```

Comment:
Please rename the variables

Reply:
I have changed the implementation of current method, instead I added dependency in pom.xml org.locationtech.jts, in order to make method be more presize( the previous considered Earth as a perfect ellipce, which is approximated the calculations). Library added helps to make more specific calculations of distance beetween two points in the map.